src/data/dfinsupp/well_founded.lean

```diff
@@ -0,0 +1,190 @@
+/-
+Copyright (c) 2022 Junyan Xu. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Junyan Xu
+-/
+import data.dfinsupp.lex
+import order.game_add
+import order.antisymmetrization
+import set_theory.ordinal.basic
+
+/-!
+# Well-foundedness of the lexicographic and product orders on `dfinsupp` and `pi`
+
+The proof is modelled on the proof of `well_founded.cut_expand`.
+-/
+
+variables {ι : Type*} {α : ι → Type*}
+
+namespace dfinsupp
+
+variables [hz : Π i, has_zero (α i)] (r : ι → ι → Prop) (s : Π i, α i → α i → Prop)
+include hz
+
+open relation prod
+
+lemma lex_fibration : fibration
+  (inv_image (game_add (dfinsupp.lex r s) (dfinsupp.lex r s)) snd)
```

Comment:
I'm not sure that `game_add` is a good name for a concept that has applications outside of combinatorial game theory. But right now, I have no better suggestion.

Reply:
Yeah the name is a bit weird especially considering the full name is `prod.game_add` (changed by @vihdzp). It's a sort of disjoint/disjunctive union like [sum.lift_rel](https://leanprover-community.github.io/mathlib_docs/data/sum/basic.html#sum.lift_rel) but for the product type; yet another viewpoint is think of it as defining the (<) operation on the product of two preorders. The (≤) operation is just given by [`prod.rprod (≤) (≤)`](https://leanprover-community.github.io/mathlib_docs/init/wf.html#prod.rprod), but (<) is defined as `a ≤ b ∧ ¬ b ≤ a` and not so straightforward. However if you take [`prod.game_add (<) (<)`](https://leanprover-community.github.io/mathlib_docs/order/game_add.html#prod.game_add) and then take its transitive closure (or just compose it with itself once), you get the (<) of the product order.

Reply:
Thanks for adding the explanation in the docstring. (I know what addition of games is, but I doubt that most mathematicians know that...)